sys/mips/mips/trap.c

```diff
@@ -870,6 +870,11 @@ trap(struct trapframe *trapframe)
 		}
 		break;
 	case T_TLB_MOD:
+		if (td->td_critnest != 0 || td->td_intr_nesting_level != 0 ||
```

Comment:
The inlining rather than using something like `data_abort` makes me sad...

Reply:
sys/mips is a sad part of the tree in general.  Most other architectures use a dedicated function for MMU exceptions particular (data_abort on arm-y things and riscv, `trap_pfault` on x86, and some other handler on powerpc.  The MIPS code is also a bit odd with how it uses the a single switch with `T_USER`.  Other architectures often use separate switches or handlers for kernel vs user traps.  Note that MIPS's machine check handler ignores machine checks that occur in userland as a result for example.  A bit of a mess to untangle and fix though, and realistically at this point I think it's more prudent to keep sys/mips on life support than to do substantial refactorings.  (Which your commit is inline with)

Reply:
I am worried though about emulation of A/D faults and whether those might (reasonably) occur on kernel addresses while holding locks.  Kernel memory in FreeBSD generally is not pageable, so most architectures just set the relevant A/D bits in PTEs or TLB entries when first programming mappings of kmem pages to avoid the overhead of pointless faults (since we don't care about tracking the A/D state of those pages).  I wonder if MIPS is careful to do the same.  If not, could we have false positives here?  I guess the general rule of thumb though is if the fault handler itself needs to look any non-spin locks (e.g. if it's going to call vm_fault()) then the WITNESS warning can give a more graceful error earlier than a deadlock in vm_fault.

Reply:
Yeah that would be my main concern. Fancy running kyua again as a stress-test? :) But such cases aren't necessary false-positives, they're still indicative of bugs in the default permissions for kernel mappings that could lead to deadlocks depending on what exactly the A/D emulation is doing.

Reply:
There is risk that an A/D emulation fault lands on a reclaimed PT page and so bails out to `vm_fault`, yes.  There have been related landmines that have gone un-noticed approximately forever, which is indicative of how rarely the PT reclaim logic actually fires (on the order of proton decay, I think); see https://reviews.freebsd.org/D24523 for example.